akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -0,0 +1,147 @@
+ * which groups a stream into vectors based on custom weight and time.
+ * The problem with that solution is each grouped vector must fit into memory before emitting to the next stage.
+ * That won't work for the use case of writing large files.
+ * The desirable behavior is to write data as they come as opposed to accumulate everything until the group slicing condition is met.
+ * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+ * This custom flow uses custom aggregator to support such use cases.
+ * Upstream inputs are continuously aggregated as they arrive.
+ * The aggregator/grouping can be terminated and emitted based on custom coditions
+ *
+ * @param seed        initiate the aggregated output with first input
+ * @param aggregate   sequentially aggregate input
+ * @param emitOnAgg   decide whether the current aggregator can be emitted, invoked after each aggregate
+ * @param emitOnTimer decide whether the current aggregator can be emitted, invoked after each timer event
+ * @param harvest     this is invoked as soon as all conditions are met before emitting to next stage
+ *                    there can be extra time between harvest and next stage receiving the emitted output
+ *                    time sensitive operations can be added here, such as closing an output channel
+ */
+class FoldWith[In, Agg, Out](
+    seed: In => Agg,
+    aggregate: (Agg, In) => Agg,
+    emitOnAgg: Agg => Boolean,
+    harvest: Agg => Out,
+    emitOnTimer: Option[(Agg => Boolean, FiniteDuration)] = None,
+    pullOnStart: Boolean = true)
+    extends GraphStage[FlowShape[In, Out]] {
+
+  emitOnTimer.foreach {
+    case (_, interval) => require(interval.gteq(1.milli), s"timer(${interval.toCoarsest}) must not be smaller than 1ms")
+  }
+
+  val in: Inlet[In] = Inlet[In](s"${this.getClass.getName}.in")
+  val out: Outlet[Out] = Outlet[Out](s"${this.getClass.getName}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      // mutable state to keep track of the aggregator status to coordinate the flow
+      // input/output handler callbacks are guaranteed to execute without concurrency
+      // https://doc.akka.io/docs/akka/current/stream/stream-customize.html#thread-safety-of-custom-operators
+      private[this] var aggregator: Agg = null.asInstanceOf[Agg]
+
+      override def preStart(): Unit = {
+        if (pullOnStart) pull(in)
+        emitOnTimer.foreach {
+          case (_, interval) => scheduleWithFixedDelay("FoldWithinIntervalTimer", interval, interval)
+        }
+      }
+
+      override protected def onTimer(timerKey: Any): Unit =
+        emitOnTimer.foreach {
+          case (isReadyOnTimer, _) => if (aggregator != null && isReadyOnTimer(aggregator)) harvestAndEmit()
```

Comment:
Shouldn't this also pull upstream if it emitted an aggregate downstream, to start on next as fast as possible?

Reply:
The problem is that emit alone does not clear the back pressure.
The downstream should be in full control in terms of pulling.
1. as long as downstream signals a pull, upstream will be pulled. No onPull from downstream means back pressure and this stage should not pull upstream.
2. upstream onPush will be continuously triggered as long as isAvailable(out). If out is not available, it also means there is back pressuring.
Imagine if we do a pull right after emit, the next input itself could trigger another emit while the previous emit is still on holding. This could potentially pile up the emit queue, due to pulling without an explicit request demand.

Reply:
Since this stage has semi-detached demand between upstream and downstream and can keep demanding while downstream backpressures, shouldn't it be like this:

In either of the cases, timer or group completed, where `harvestAndEmit` did in fact signal trigger an emit, then there should be a pull to start the next aggregate as fast as possible, if downstream still backpressures once that group is completed (or the next timer hits), that should just mean the complete aggregate stays as is until there is demand and it can be pushed downstream (with no further demand is sent upstream until that time).

Reply:
Could it use the on-emit-complete-callback overload for that pull then to avoid piling up?

Reply:
Just did another generalization to decouple the upstream and downstream with internal buffer.
It also handles the special case of buffer=1 more naturally in terms of eager pull with timely emit.
The updated test case should demonstrate the expected behaviors in detail.